cms_coding_rules.md

```diff
@@ -117,6 +117,7 @@ If necessary to create a unique name, one can add the directory name:
 #### Python
 13. A `_cfi` file should contain only the definition of one module, and possible Modifier ("era") customizations on it. The module label should be the same as the `_cfi` file name.
 14. The `_cfi` file should be left to be generated automatically with the `fillDescriptions()`. When Modifier customizations are needed, the auto-generated label should have e.g. "Default" postfix and be imported+cloned to the desired name.
+  Whenever the plugin needs to be included in a HLT configuration, it is mandatory (for confDB parsing purposes) that a default value is provided for all the configurable parameters; for the HLT case the usage of `addDefault()` is discouraged. In general it is preferable to use `addWithDefaultLabel()`, such that the auto-generated cfi file bears the same name as the plugin, starting with a lower case character.
 15. A module/Task/Sequence/Path with a given name should be defined in exactly one `_cfi` or `_cff` file.
```

Comment:
Hmm, the `default` in the first half/second half of the first sentence is a different default. I suggest:

Whenever the plugin needs to be included in a HLT configuration, it is mandatory (for confDB parsing purposes) that a default value is provided for all the top-level parameters. This is typically done by coding a `fillDescription(...)` method and use it to write out a cfi file with ```addWithDefaultLabel(...)``` for ConfDb parsing. For the HLT case the usage of `addDefault(...)` is discouraged as it does not write out such a cfi file.

Sorry to be so overly picky, but better spell it out in detail.

Reply:
Well, given there are frequent problems with this issue, I think the level of detail is warranted.

Reply:
I'd nitpick to change the
```md
cfi
```
with
```md
`_cfi`
```
to be consistent with the surrounding formatting.

For the "This is typically done by ... all these parameters" sentence, how about something along
```md
This should be done in the `fillDescriptions()` method by specifying all these top-level parameters with a default value, and using the `addWithDefaultLabel()` method to write out a `_cfi` file containing values for all these parameters.
```
(The `fillDescriptions()` is already introduced, and I understand "should be done" to communicate higher level of requirement than "typically done"; also the first sentence of 14. uses "should be")

Is the `The method `addWithDefaultLabel()` creates ... in ConfDb parsing." sentence really necessary? Could it be replaced with a link to https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Automatic_module_labels_from_plu ? (e.g. making the first occurrence of `addWithDefaultLabel()` to be the link) Further explanation of the behavior (like the first letter being lower case) could be easily added there.

Reply:
I applied the suggestions in the last push, and removed:

```
 The method `addWithDefaultLabel()` creates a cfi file with the same name as the plugin (starting with a lower-case letter) which is used in ConfDb parsing. For the HLT case the usage of the method `addDefault()` is thus discouraged as it does not write out any cfi file.
 ```
 Can it be added to https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#Automatic_module_labels_from_plu ?